pylint/checkers/variables.py

```diff
@@ -976,267 +989,261 @@ def visit_assignname(self, node: nodes.AssignName) -> None:
+                if isinstance(frame, nodes.ClassDef) and node.name in frame.locals:
+                    if isinstance(node.parent, nodes.Arguments):
+                        if stmt.fromlineno <= defstmt.fromlineno:
+                            # Doing the following is fine:
+                            #   class A:
+                            #      x = 42
+                            #      y = lambda attr=x: attr
                             self.add_message(
-                                "undefined-variable", node=node, args=node.name
+                                "used-before-assignment",
+                                args=node.name,
+                                node=node,
                             )
-
-            current_consumer.mark_as_consumed(node.name, found_nodes)
-            # check it's not a loop variable used outside the loop
-            self._loopvar_name(node)
-            break
-        else:
-            # we have not found the name, if it isn't a builtin, that's an
-            # undefined name !
-            if undefined_variable_is_enabled and not (
-                node.name in nodes.Module.scope_attrs
-                or utils.is_builtin(node.name)
-                or node.name in self.config.additional_builtins
-                or (
-                    node.name == "__class__"
-                    and isinstance(frame, nodes.FunctionDef)
-                    and frame.is_method()
-                )
-            ):
-                if not utils.node_ignores_exception(node, NameError):
+                    else:
+                        self.add_message(
+                            "undefined-variable", args=node.name, node=node
+                        )
+                        return (VariableVisitConsumerAction.CONSUME, found_nodes)
+                elif current_consumer.scope_type == "lambda":
+                    self.add_message("undefined-varibale", args=node.name, node=node)
+                    return (VariableVisitConsumerAction.CONSUME, found_nodes)
+
+        elif self._is_only_type_assignment(node, defstmt):
+            self.add_message("undefined-variable", args=node.name, node=node)
+            return (VariableVisitConsumerAction.CONSUME, found_nodes)
+
+        elif isinstance(defstmt, nodes.ClassDef):
+            is_first_level_ref = self._is_first_level_self_reference(node, defstmt)
+            if is_first_level_ref == 2:
+                self.add_message("used-before-assignment", node=node, args=node.name)
+            if is_first_level_ref:
+                return (VariableVisitConsumerAction.RETURN, None)
+
+        elif isinstance(defnode, nodes.NamedExpr):
+            if isinstance(defnode.parent, nodes.IfExp):
+                if self._is_never_evaluated(defnode, defnode.parent):
                     self.add_message("undefined-variable", args=node.name, node=node)
+                    return (VariableVisitConsumerAction.CONSUME, found_nodes)
```

Comment:
```suggestion
        elif (
            isinstance(defnode, nodes.NamedExpr)
            and isinstance(defnode.parent, nodes.IfExp)
            and self._is_never_evaluated(defnode, defnode.parent)
        ):
            self.add_message("undefined-variable", args=node.name, node=node)
            return (VariableVisitConsumerAction.CONSUME, found_nodes)
```

Reply:
I did this so we can easily change and add checks for when there is a `NamedExpr`. I prefer to keep these top if's as `elif isinstance(defnode, nodes.XXX):` so we can easily add things later on.